From dbefc642b3bd0cafe599fcd18c6c88cb672397ee Mon Sep 17 00:00:00 2001
From: Grimmauld <Grimmauld@grimmauld.de>
Date: Thu, 26 Jun 2025 09:56:50 +0200
Subject: [PATCH 1/3] auditd: Avoid blocking on open syscall

`open` might block on FIFO files or some character/block devices such as /dev/tty.
Checking `stat` based on path introduces a TOCTOU issue.

diff --git a/audisp/audispd-pconfig.c b/audisp/audispd-pconfig.c
index d334de78..1479cbc9 100644
--- a/audisp/audispd-pconfig.c
+++ b/audisp/audispd-pconfig.c
@@ -134,7 +134,7 @@ void clear_pconfig(plugin_conf_t *config)
 	config->restart_cnt = 0;
 }
 
-int load_pconfig(plugin_conf_t *config, char *file)
+int load_pconfig(plugin_conf_t *config, int dirfd, char *file)
 {
 	int fd, rc, mode, lineno = 1;
 	struct stat st;
@@ -143,9 +143,12 @@ int load_pconfig(plugin_conf_t *config, char *file)
 
 	clear_pconfig(config);
 
-	/* open the file */
-	mode = O_RDONLY;
-	rc = open(file, mode);
+	/* O_NONBLOCK avoids blocking when opening a FIFO file accidentially.
+	 * It does however block if someone symlinks /dev/ttyX into the plugin directory.
+	 * We do not pass O_NOFOLLOW, which allows for symlinked configs.
+	 */
+	mode = O_RDONLY | O_NONBLOCK;
+	rc = openat(dirfd, file, mode);
 	if (rc < 0) {
 		if (errno != ENOENT) {
 			audit_msg(LOG_ERR, "Error opening %s (%s)", file,
@@ -159,7 +162,7 @@ int load_pconfig(plugin_conf_t *config, char *file)
 	fd = rc;
 
 	/* check the file's permissions: owned by root, not world writable,
-	 * not symlink.
+	 * pointing to regular file.
 	 */
 	if (fstat(fd, &st) < 0) {
 		audit_msg(LOG_ERR, "Error fstat'ing config file (%s)",
@@ -179,6 +182,7 @@ int load_pconfig(plugin_conf_t *config, char *file)
 		close(fd);
 		return 1;
 	}
+	// this checks if the actual file is a regular file. The initial open might have followed a symlink, which is acceptable.
 	if (!S_ISREG(st.st_mode)) {
 		audit_msg(LOG_ERR, "Error - %s is not a regular file",
 			file);
@@ -186,6 +190,13 @@ int load_pconfig(plugin_conf_t *config, char *file)
 		return 1;
 	}
 
+	if (fcntl(fd, F_SETFL, mode & (~O_NONBLOCK)) < 0) {
+		audit_msg(LOG_ERR, "Error - Failed to remove nonblock flag for %s",
+			file);
+		close(fd);
+		return 1;
+	}
+
 	/* it's ok, read line by line */
 	f = fdopen(fd, "rm");
 	if (f == NULL) {
diff --git a/audisp/audispd-pconfig.h b/audisp/audispd-pconfig.h
index 3f6c2961..2d89d116 100644
--- a/audisp/audispd-pconfig.h
+++ b/audisp/audispd-pconfig.h
@@ -50,7 +50,7 @@ typedef struct plugin_conf
 } plugin_conf_t;
 
 void clear_pconfig(plugin_conf_t *config);
-int  load_pconfig(plugin_conf_t *config, char *file);
+int  load_pconfig(plugin_conf_t *config, int dirfd, char *file);
 void free_pconfig(plugin_conf_t *config);
 
 #endif
diff --git a/audisp/audispd.c b/audisp/audispd.c
index 2113d4f9..0023c7b5 100644
--- a/audisp/audispd.c
+++ b/audisp/audispd.c
@@ -111,11 +111,16 @@ static void load_plugin_conf(conf_llist *plugin)
 	/* read configs */
 	d = opendir(daemon_config.plugin_dir);
 	if (d) {
+		int dfd = dirfd(d);
+		if (dfd < 0) {
+			closedir(d);
+			return;
+		}
+
 		struct dirent *e;
 
 		while ((e = readdir(d))) {
 			plugin_conf_t config;
-			char fname[PATH_MAX];
 			const char *ext, *reason = NULL;
 
 			if (e->d_type != DT_REG)
@@ -132,11 +137,8 @@ static void load_plugin_conf(conf_llist *plugin)
 				continue;
 			}
 
-			snprintf(fname, sizeof(fname), "%s/%s",
-				daemon_config.plugin_dir, e->d_name);
-
 			clear_pconfig(&config);
-			if (load_pconfig(&config, fname) == 0) {
+			if (load_pconfig(&config, dfd, e->d_name) == 0) {
 				/* Push onto config list only if active */
 				if (config.active == A_YES)
 					plist_append(plugin, &config);
-- 
2.49.0

